alf/environments/simple/noisy_array.py

```diff
@@ -0,0 +1,173 @@
+        return self._obs
+
+    def step(self, action):
+        # auto_reset will be called by wrappers
+        self._action = action
+        self._obs, r = self._gen_observation(act=self._action)
+        return self._obs, r, self._game_over, {}
+
+    def render(self, mode="human", close=False):
+        # first convert obs to an RGB array
+        obs = np.copy(1 - self._obs)
+        obs[self._K:] *= 0.5  # turn the noise portion to gray
+        obs *= 255
+        obs = obs.astype("uint8")
+
+        grid_size = 16
+        length = obs.shape[0]
+        rgb_array = np.expand_dims(obs, axis=0)
+        rgb_array = cv2.resize(
+            rgb_array,
+            dsize=(length * grid_size, grid_size),
+            interpolation=cv2.INTER_NEAREST)
+        rgb_array = cv2.cvtColor(rgb_array, cv2.COLOR_GRAY2RGB)
+
+        if mode == "rgb_array":
+            return rgb_array
+        else:
+            if self._action is not None:
+                print("action: ", self._action)
+            cv2.imshow("NoisyArray", rgb_array)
+            cv2.waitKey(100)
+
+    def _gen_observation(self, act):
+        movement = act - 1
+        # If the current position is beyond the right boundary, put the agent
+        # back to the left
+        self._position = max(self._position + movement, 0)
+        self._position %= self._K
+
+        self._game_over = (self._position == self._K - 1)
+
+        # step type
+        reward = 1 if self._game_over else 0
+
+        if act == self.FIRE:
+            self._noise_vector = np.random.randint(2, size=self._M)
+
+        position_array = np.zeros(self._K, dtype=np.float32)
+        position_array[self._position] = 1
+
+        observation = np.concatenate((position_array,
+                                      self._noise_vector.astype(np.float32)))
+        return observation, reward
+
+
+class NoisyArrayTest(parameterized.TestCase, unittest.TestCase):
+    @parameterized.parameters((2, 3), (100, 100))
+    def test_noisy_array_environment(self, K, M):
+        array = NoisyArray(K, M)
+        array.reset()
+        for _ in range(2 * K):
+            done = array.step(NoisyArray.RIGHT)[2]
+        self.assertTrue(done)
+
+        array.reset()
+        array.step(NoisyArray.LEFT)  # cannot go beyond the left boundary
+        self.assertEqual(array._position, 0)
+
+        array.step(NoisyArray.RIGHT)
+        array.reset()
+
+        game_ends = 0
+        total_rewards = 0
+        done = False
+        for _ in range(4 * K + 1):
+            if done:
+                array.reset()
+                done = False
+            else:
+                _, r, done, _ = array.step(NoisyArray.RIGHT)
+                total_rewards += r
+                game_ends += int(done)
+
+        self.assertEqual(game_ends, 2)
+        self.assertEqual(total_rewards, 2)
+        self.assertEqual(r, 1.0)
+        self.assertTrue(done)
+
+
+if __name__ == "__main__":
+    unittest.main()
```

Comment:
Should put the test to a separate file ending with _test.py so that CI will run it automatically.

Reply:
Done.